core/src/main/java/org/infinispan/util/concurrent/locks/impl/DefaultLockManager.java

```diff
@@ -173,6 +174,13 @@ public void unlockAll(Collection<?> keys, Object lockOwner) {
    public void unlockAll(InvocationContext context) {
       unlockAll(context.getLockedKeys(), context.getLockOwner());
       context.clearLockedKeys();
+      if (context instanceof TxInvocationContext<?>) {
+         // this may be on overkill but if the TM's Transaction Reaper aborts a transaction and a lock is not acquired
+         // (i.e it is in WAITING state) when the RollbackCommand is executed, the lock is not released.
+         // In other words, WAITING state moves to ACQUIRED state and the RollbackCommand is never executed again
+         // leaving the lock acquired forever
+         unlockAll(((TxInvocationContext<?>) context).getAffectedKeys(), context.getLockOwner());
```

Comment:
This seems wrong to me. Why not just mark the keys as locked keys collection before attempting to lock them, so that the unlockAll(context.getLockedKeys(), context.getLockOwner()) call releases everything?

Reply:
retries are problematic. if don't try to acquire the lock if the key is in the `lockedKeys` set: https://github.com/infinispan/infinispan/blob/ad9fcce5fb94166b49c1e4ebbf3d0c43b02579d9/core/src/main/java/org/infinispan/interceptors/locking/AbstractTxLockingInterceptor.java#L89

what I can do is change `lockedKeys` to a Map to signal if the key is locked or "waiting" for the lock

Reply:
No, I wouldn't want to add one more map access to change the locked status on the successful path.

I don't think retries would be a problem, because we never decide to retry a lock command while it's waiting for a lock. Either acquire the lock and then retry, or we retry before trying to acquire the lock.

If you decide to keep this approach, though, please change the wording in the comment, because releasing all affected keys is not "overkill" when you don't have a choice :)